### DISCUSSION TITLE
Best way to inject beans into instance fields

### DISCUSSION THREAD

**Discussion by rsenden** [micronaut-projects/micronaut-core#6375] [Q&A] [answered]
Being new to Micronaut, I'm trying to find the best/most concise way to inject beans into instance fields, which works both in a regular JVM and GraalVM native image.

Below are some of the approaches that I've tried so far (note that I'm also using some Lombok annotations). All of these approaches work fine when running in a normal JVM, but most of these fail when running as a GraalVM native image.

Any suggestions or alternative approaches that I didn't think of yet?

## `@Inject` on private field (not working)
Fails at runtime with `Error setting field value: No field 'myBean' found for type`.
```
@Singleton public MyClass {
  @Getter @Inject private MyBean myBean;
}
```

## `@Inject` and `@ReflectiveAccess` on private field (not working)
Fails at runtime with `Error setting field value: No field 'myBean' found for type`, despite several suggestions in related posts that this should work.
```
@Singleton public MyClass {
  @Getter @Inject @ReflectiveAccess private MyBean myBean;
}
```

## `@Inject` on package-private field (not working)
Fails at runtime with `Error setting field value: No field 'myBean' found for type`, despite several suggestions in related posts that this should work with package-protected fields.
```
@Singleton public MyClass {
  @Getter @Inject MyBean myBean;
}
```

## Lombok-generated setter with `@Inject` annotation (not working)
Bean is not being injected when running in GraalVM, and in general this suffers from the same disadvantage as setter-based injection as described below.
```
@Singleton public MyClass {
  @Getter @Setter(onMethod_= {@Inject}) private MyBean myBean;
}
```

## Setter-base injection (working)
Main disadvantage is that injection is not guaranteed; I think the setter will not be called if no qualifying bean is available (rather than having Micronaut throw an error like with field-based or constructor-based injection).
```java
@Singleton public MyClass {
  private MyBean myBean;
  @Inject public void setMyBean(MyBean myBean) { this.myBean = myBean; }
}
```

## Constructor-based injection (working)
This is probably the cleanest approach, but also the least concise option especially with class hierarchies, having to implement constructor chaining if some superclass requires some bean provided by Micronaut DI.
```java
@Singleton public MyClass {
  private final MyBean myBean;
  @Inject public MyClass(MyBean myBean) { this.myBean = myBean; }
}
```

## Gradle configuration
In case it matters, following lists the relevant Gradle configuration sections:

```
plugins {
	id("com.github.johnrengelman.shadow") version "7.0.0"
	id("io.micronaut.application") version "2.0.6" apply false
	id("io.micronaut.library") version "2.0.6" apply false
}

apply plugin: "io.micronaut.application"
subprojects {
	apply plugin: "io.micronaut.library"
}

allprojects {
	dependencies {
		// Lombok dependency & annotation processor
		compileOnly 'org.projectlombok:lombok:1.16.20'
		annotationProcessor "org.projectlombok:lombok:1.16.20"
		
		// Picocli dependency and annotation processor
		annotationProcessor("info.picocli:picocli-codegen")
		implementation("info.picocli:picocli")
		
		// Micronaut dependencies
		implementation("io.micronaut:micronaut-runtime")
		implementation("io.micronaut.picocli:micronaut-picocli")
		implementation("javax.annotation:javax.annotation-api")
	}
}
```

**Comment by graemerocher** [comment, marked as answer]
you need the `micronaut-graal` processor to make the other use cases work as reflection data has to be registered and your build is missing that processor

**Reply by rsenden** [reply]
Thanks for the tip, I figured I must be missing something. I have now updated my Gradle configuration to include the annotation processor. Unfortunately though I'm still seeing the same errors when using either `@Getter @Inject MyBean myBean` or `@Getter @Inject @ReflectiveAccess MyBean myBean` (both package-private, with or without `@ReflectiveAccess`).

Could this be related to having those injected fields on abstract classes? The exception that I'm currently getting refers to an abstract class that contains the `@Inject @ReflectiveAccess` annotations:

```
Caused by: java.lang.NoSuchFieldError: No field 'myField' found for type: com.mypackage.AbstractMyClass
	at io.micronaut.core.reflect.ReflectionUtils.lambda$getRequiredField$2(ReflectionUtils.java:290)
	at java.util.Optional.orElseThrow(Optional.java:408)
	at io.micronaut.core.reflect.ReflectionUtils.getRequiredField(ReflectionUtils.java:290)
```

For reference, following is the relevant dependencies section after adding the annotation processor:
```
		// Lombok dependency & annotation processor
		compileOnly 'org.projectlombok:lombok:1.16.20'
		annotationProcessor "org.projectlombok:lombok:1.16.20"
		
		// Picocli dependency and annotation processor
		annotationProcessor("info.picocli:picocli-codegen")
		implementation("info.picocli:picocli")
		
		annotationProcessor("io.micronaut:micronaut-inject-java")
		
		// Micronaut dependencies
		implementation("io.micronaut:micronaut-runtime")
		implementation("io.micronaut.picocli:micronaut-picocli")
		implementation("javax.annotation:javax.annotation-api")

```

**Reply by rsenden** [reply]
My bad, I added the `micronaut-inject-java` annotation processor but not `micronaut-graal`, will update on the results once my build finishes

**Reply by rsenden** [reply]
I'll need to do some more experimenting on which approaches actually work now, but at least `@Getter @Inject @ReflectiveAccess MyBean myBean;` seems to be working fine now.
